File: fastapi/app/schemas/team.py
from __future__ import annotations

from datetime import datetime
from typing import List

from pydantic import BaseModel, EmailStr, Field


class TeamMemberDetail(BaseModel):
    user_id: int
    email: EmailStr
    role: str
    status: str
    invited_by_id: int | None = None
    joined_at: datetime


class TeamBase(BaseModel):
    name: str = Field(min_length=1, max_length=127)


class TeamCreate(TeamBase):
    pass


class TeamDetail(TeamBase):
    id: int
    owner_id: int
    created_at: datetime
    updated_at: datetime
    members: List[TeamMemberDetail] = Field(default_factory=list)

    class Config:
        from_attributes = True


class TeamInviteRequest(BaseModel):
    email: EmailStr
    role: str = Field(default="member", max_length=32)


class TeamMemberUpdate(BaseModel):
    role: str | None = Field(default=None, max_length=32)
    status: str | None = None

File: nextjs/src/lib/team.ts
"use client";

import api from "@/lib/api";
import { API_ENDPOINTS } from "@/lib/config";

export type TeamMember = {
  user_id: number;
  email: string;
  role: string;
  status: string;
  invited_by_id: number | null;
  joined_at: string;
};

export type Team = {
  id: number;
  name: string;
  owner_id: number;
  created_at: string;
  updated_at: string;
  members: TeamMember[];
};

export type CreateTeamPayload = {
  name: string;
};

export type InviteMemberPayload = {
  teamId: number;
  email: string;
  role: string;
};

export type UpdateMemberPayload = {
  teamId: number;
  memberId: number;
  role?: string;
  status?: string;
};

export const getTeams = () => api.get<Team[]>(API_ENDPOINTS.teams.root);

export const createTeam = (payload: CreateTeamPayload) =>
  api.post<Team>(API_ENDPOINTS.teams.root, payload);

export const inviteMember = ({ teamId, ...payload }: InviteMemberPayload) =>
  api.post<TeamMember>(API_ENDPOINTS.teams.invite(teamId), payload);

export const updateMember = ({ teamId, memberId, ...payload }: UpdateMemberPayload) =>
  api.patch<TeamMember>(API_ENDPOINTS.teams.member(teamId, memberId), payload);
